### PR TITLE
SEO-577: 404 response on categories without members or media

### PR DESCRIPTION
**Pull request by saipetch**
SEO:
- 404 response on categories without members or media
- exclude category pages without members or media on SitemapXml

### CODE REVIEW

File: extensions/wikia/MercuryApi/MercuryApiController.class.php [hunk 1]
```diff
@@ -360,17 +360,19 @@ public function getPage() {
 					case NS_CATEGORY:
```
Comment (WikiaTech) [inline extensions/wikia/MercuryApi/MercuryApiController.class.php:360]: MAJOR Reduce this "switch/case" number of lines from 19 to at most 10, for example by extracting code into function. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1151)

File: extensions/wikia/MercuryApi/MercuryApiController.class.php [hunk 2]
```diff
@@ -360,17 +360,19 @@ public function getPage() {
...
-							$data['details'] = MercuryApiArticleHandler::getArticleDetails( $article );
-							$data['article'] = MercuryApiArticleHandler::getArticleJson( $this->request, $article );
-
-							// Remove namespace prefix from displayTitle, so it can be consistent with title
-							// Prefix shows only if page doesn't have {{DISPLAYTITLE:title} in it's markup
-							$data['article']['displayTitle'] = Title::newFromText($data['article']['displayTitle'])->getText();
-						} elseif ( !empty( $data['nsSpecificContent']['members']['sections'] ) ) {
-							$data['details'] = MercuryApiCategoryHandler::getCategoryMockedDetails( $title );
+						if ( !empty( $data['nsSpecificContent']['members']['sections'] ) ) {
+							if ( MercuryApiCategoryHandler::hasArticle( $this->request, $article ) ) {
```
Comment (WikiaTech) [inline extensions/wikia/MercuryApi/MercuryApiController.class.php:364]: MAJOR Refactor this code to not nest more than 4 "if", "for", "while", "switch" and "try" statements. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS134)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 21 issues
* MAJOR 12 major
* MINOR 8 minor
* INFO 1 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [MercuryApiController.class.php#L379](https://github.com/Wikia/app/blob/78f213cbf2c923d1db3fee677d27a2455b6ba4b5/extensions/wikia/MercuryApi/MercuryApiController.class.php#L379): Reduce this "switch/case" number of lines from 16 to at most 10, for example by extracting code into function. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1151)
1. MAJOR [MercuryApiController.class.php#L381](https://github.com/Wikia/app/blob/78f213cbf2c923d1db3fee677d27a2455b6ba4b5/extensions/wikia/MercuryApi/MercuryApiController.class.php#L381): Refactor this code to not nest more than 4 "if", "for", "while", "switch" and "try" statements. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS134)
1. MAJOR [MercuryApiController.class.php#L517](https://github.com/Wikia/app/blob/78f213cbf2c923d1db3fee677d27a2455b6ba4b5/extensions/wikia/MercuryApi/MercuryApiController.class.php#L517): Remove this commented out code. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS125)
1. MAJOR [SEOTweaksHooksHelper.class.php#L40](https://github.com/Wikia/app/blob/78f213cbf2c923d1db3fee677d27a2455b6ba4b5/extensions/wikia/SEOTweaks/SEOTweaksHooksHelper.class.php#L40): Remove the unused function parameter "$article". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1172)
1. MAJOR [SEOTweaksHooksHelper.class.php#L44](https://github.com/Wikia/app/blob/78f213cbf2c923d1db3fee677d27a2455b6ba4b5/extensions/wikia/SEOTweaks/SEOTweaksHooksHelper.class.php#L44): Add a "case default" clause to this "switch" statement. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS131)
1. MAJOR [SEOTweaksHooksHelper.class.php#L73](https://github.com/Wikia/app/blob/78f213cbf2c923d1db3fee677d27a2455b6ba4b5/extensions/wikia/SEOTweaks/SEOTweaksHooksHelper.class.php#L73): Remove the unused function parameter "$html". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1172)
1. MAJOR [SEOTweaksHooksHelper.class.php#L107](https://github.com/Wikia/app/blob/78f213cbf2c923d1db3fee677d27a2455b6ba4b5/extensions/wikia/SEOTweaks/SEOTweaksHooksHelper.class.php#L107): Remove the unused function parameters "$parser, $params, $time, $descQuery, $options". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1172)
1. MAJOR [SEOTweaksHooksHelper.class.php#L125](https://github.com/Wikia/app/blob/78f213cbf2c923d1db3fee677d27a2455b6ba4b5/extensions/wikia/SEOTweaks/SEOTweaksHooksHelper.class.php#L125): Remove the unused function parameter "$pcache". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1172)
1. MAJOR [SitemapXmlController.class.php#L191](https://github.com/Wikia/app/blob/78f213cbf2c923d1db3fee677d27a2455b6ba4b5/extensions/wikia/SitemapXml/SitemapXmlController.class.php#L191): Reduce the number of returns of this function 5, down to the maximum allowed 3. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1142)
1. MAJOR [SitemapXmlController.class.php#L220](https://github.com/Wikia/app/blob/78f213cbf2c923d1db3fee677d27a2455b6ba4b5/extensions/wikia/SitemapXml/SitemapXmlController.class.php#L220): Reduce the number of returns of this function 4, down to the maximum allowed 3. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1142)

**Comment by WikiaTech** [comment]
SonarQube analysis reported 18 issues
* MAJOR 10 major
* MINOR 7 minor
* INFO 1 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [MercuryApiController.class.php#L379](https://github.com/Wikia/app/blob/fab2c36208c93d3d28a9a1bed8aca86348a1557d/extensions/wikia/MercuryApi/MercuryApiController.class.php#L379): Reduce this "switch/case" number of lines from 16 to at most 10, for example by extracting code into function. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1151)
1. MAJOR [MercuryApiController.class.php#L381](https://github.com/Wikia/app/blob/fab2c36208c93d3d28a9a1bed8aca86348a1557d/extensions/wikia/MercuryApi/MercuryApiController.class.php#L381): Refactor this code to not nest more than 4 "if", "for", "while", "switch" and "try" statements. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS134)
1. MAJOR [MercuryApiController.class.php#L517](https://github.com/Wikia/app/blob/fab2c36208c93d3d28a9a1bed8aca86348a1557d/extensions/wikia/MercuryApi/MercuryApiController.class.php#L517): Remove this commented out code. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS125)
1. MAJOR [SEOTweaksHooksHelper.class.php#L38](https://github.com/Wikia/app/blob/fab2c36208c93d3d28a9a1bed8aca86348a1557d/extensions/wikia/SEOTweaks/SEOTweaksHooksHelper.class.php#L38): Remove the unused function parameter "$html". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1172)
1. MAJOR [SEOTweaksHooksHelper.class.php#L72](https://github.com/Wikia/app/blob/fab2c36208c93d3d28a9a1bed8aca86348a1557d/extensions/wikia/SEOTweaks/SEOTweaksHooksHelper.class.php#L72): Remove the unused function parameters "$parser, $params, $time, $descQuery, $options". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1172)
1. MAJOR [SEOTweaksHooksHelper.class.php#L90](https://github.com/Wikia/app/blob/fab2c36208c93d3d28a9a1bed8aca86348a1557d/extensions/wikia/SEOTweaks/SEOTweaksHooksHelper.class.php#L90): Remove the unused function parameter "$pcache". [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1172)
1. MAJOR [SitemapXmlController.class.php#L191](https://github.com/Wikia/app/blob/fab2c36208c93d3d28a9a1bed8aca86348a1557d/extensions/wikia/SitemapXml/SitemapXmlController.class.php#L191): Reduce the number of returns of this function 5, down to the maximum allowed 3. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1142)
1. MAJOR [SitemapXmlController.class.php#L220](https://github.com/Wikia/app/blob/fab2c36208c93d3d28a9a1bed8aca86348a1557d/extensions/wikia/SitemapXml/SitemapXmlController.class.php#L220): Reduce the number of returns of this function 4, down to the maximum allowed 3. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1142)
1. MINOR [MercuryApiController.class.php#L10](https://github.com/Wikia/app/blob/fab2c36208c93d3d28a9a1bed8aca86348a1557d/extensions/wikia/MercuryApi/MercuryApiController.class.php#L10): Define a constant instead of duplicating this literal "title" 3 times. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1192)
1. MINOR [MercuryApiController.class.php#L316](https://github.com/Wikia/app/blob/fab2c36208c93d3d28a9a1bed8aca86348a1557d/extensions/wikia/MercuryApi/MercuryApiController.class.php#L316): Define a constant instead of duplicating this literal "article" 5 times. [rule](https://sonarqube.wikia-inc.com/coding_rules#rule_key=php%3AS1192)

**Comment by gabrys** [comment]
LGTM

**Comment by rogatty** [comment]
@saipetch @gabrys do I understand this code correctly?

I think that for categories with content but without members:
- Oasis will return 404 status code and display the content
- Mercury will return 404 with "Not found" message instead of the content

If this is true - are we comfortable with this inconsistency?

**Pull request merged**